### ISSUE TITLE
[Proposal] Don’t parse Actions steps on Gitea side; instead, let the runner report them

### ISSUE THREAD

**Issue by wolfogre** [go-gitea/gitea#24604]
## Background

Gitea will parse the steps of a job from the workflow file even before running the job. It works like:

- Gitea reads the workflow file, so it knows how many steps there are and what their names are.
- The information of steps will be stored in the DB.
- When a runner runs the jobs, it doesn't need to report the number and name of steps. Instead, it reports the status and log of each step with a simple index, such as "step 5 was successful.”

I have to admit, it is a design mistake. The point is that **it's impossible to know the real steps based on the static content of the file**.

For example:

- Some actions may has Pre and Post step.
![image](https://github.com/go-gitea/gitea/assets/9418365/dc488983-f27a-4d2e-b0f2-7b577f1fe3f9)
- A job with service will have Initialize containers and Stop containers steps
![image](https://github.com/go-gitea/gitea/assets/9418365/0a51dbff-0311-4936-93ec-d002eb4528b0)
- When a job reuse another workflow, the real steps is in another file.
![image](https://github.com/go-gitea/gitea/assets/9418365/7c944bd5-d26c-4827-993b-4540e432bcfe)

At present, the problem is not really serious. You can find the logs of the missing steps in `Set up job` and `Complete job`. It's just a little inconvenient, Gitea doesn't lose the logs.

<img width="499" alt="image" src="https://github.com/go-gitea/gitea/assets/9418365/4423a2b6-bccf-47bf-94d5-58a4b2d54b9f">

## Solution

The right way could be let the runner report the steps when running them.

I have to say, it will be a huge job that requires refactoring both Gitea and act runner. Many details need to be worked out. For example, the runner should report all steps before executing them, so users can see steps waiting to be executed:

<img width="428" alt="image" src="https://github.com/go-gitea/gitea/assets/9418365/90cc42f9-f0c9-490e-a90b-3f566603ebdc">

**Comment by yekanchi** [comment]
this is also related to this issue on the runner side:
https://gitea.com/gitea/act_runner/issues/444#issuecomment-809931

**Comment by sahil87** [comment]
Instead of tackling this by taking the long route of refactoring gitea and act runner interactions, which might take a very long time or never happen, maybe we can just handle this in the frontend with the same logs? 

The UI could be a list of steps within another list of steps, generated using these recurring separators already present in the logs:

![Image](https://github.com/user-attachments/assets/a0be3748-1341-4740-b409-0f0535f2c9ce)

Even if we don't know the number of steps before hand, and get to see only those steps which have already run, it is still a huge improvement over the current situation.
Because I think the problem is definitely more than serious - its enough to keep some people searching for alternatives to gitea act runners 🤷 .

**Comment by maantje** [comment]
@sahil87, it's possible to make some improvements from the frontend.

In our instance, we added a script to the footer in `/templates/custom/footer.tmpl`. This script adds a 🧹 icon next to the action settings icon. To use it, expand the `Set up job` step and click the 🧹 icon. It will then automatically scroll through the output and omit any irrelevant logs. If an error occurs, it will stop at the error. If the job has already completed, clicking the 🧹 icon will clean up the log and scroll directly to the error.

You can try it out on a job of your own by simply pasting the contents of the `eventListener` into the browser console while on the job page.

```html
<script>
window.addEventListener('load', function () {
  const right = document.getElementsByClassName('job-info-header-right')[0];
  let interval = null
  if (right) {
      const div = document.createElement('div')

      div.classList.add('ui', 'top', 'right', 'dropdown', 'custom', 'jump', 'item')
      div.innerText = "🧹"

      const cleanUp = () => {
          window.scrollTo(0, document.body.scrollHeight);

          [...document.getElementsByClassName("log-msg")].forEach((el) => {
            for (let prefix of [
              "evaluating",
              "expression",
              "Writing",
              "Extracting",
              "Merged container",
              "Custom container",
              "[command]",
              "Exec command '[",
              "Working directory '/workspace",
              "  🐳  docker",
              "executing remote job container",
              "type=remote-action ",
              "Stripping prefix:",
              "run post step for",
              "::set-output::",
              "::add-matcher::",
              "/var/run/act/actions",
            ]) {
              if (el.innerText.startsWith(prefix)) {
                el.parentElement.remove();
                return;
              }
            }

            const match = el.innerText.split('\\n\\n');

            if (match.length === 3) {
             el.innerText = match[1]
            }

            if (el.innerText.startsWith("  ❌")) {
              if (interval) {
                 clearInterval(interval)
              }
              el.scrollIntoView(false);
            }

            if (el.innerText === "🏁  Job succeeded" && interval) {
                clearInterval(interval)
            }
          });
      }
      div.onclick = () => {
        if (interval) {
           clearInterval(interval)
           interval = null
           return
        }

        cleanUp()
        interval = setInterval(cleanUp, 100)
      }

      right.prepend(div)
  }
})
</script>

```

**Comment by sahil87** [comment]
@maantje this is really cool!

**Comment by silverwind** [comment]
Is https://github.com/go-gitea/gitea/issues/26736 a duplicate? I've posted a analysis in https://github.com/go-gitea/gitea/issues/26736#issuecomment-4128992610. Fixing this would require changes in 3 repos (as of today), so a hard fix unfortunately.

@lunny I would recommend merging both `act` and `act_runner` into the `gitea` repo so such cross-repo stuff can be done without complicated (of not impossible) coordination between 3 repos.

**Comment by bircni** [comment]
Related: #36983

**Comment by lunny** [comment]
> Is [#26736](https://github.com/go-gitea/gitea/issues/26736) a duplicate? I've posted a analysis in [#26736 (comment)](https://github.com/go-gitea/gitea/issues/26736#issuecomment-4128992610). Fixing this would require changes in 3 repos (as of today), so a hard fix unfortunately.
> 
> [@lunny](https://github.com/lunny) I would recommend merging both `act` and `act_runner` into the `gitea` repo so such cross-repo stuff can be done without complicated (of not impossible) coordination between 3 repos.

I think we can do it step by step. First of all, merge `act` and `act_runner`. But this is out of topic. The proposal is to suggest Gitea should not parse all the steps but always accept a step declare from the runner.

**Comment by silverwind** [comment]
Yeah do that merge first, but long term I think we need all in the same repo to effectively execute changes that affect all 3.

**Comment by Zettat123** [comment]
> Is [#26736](https://github.com/go-gitea/gitea/issues/26736) a duplicate? I've posted a analysis in [#26736 (comment)](https://github.com/go-gitea/gitea/issues/26736#issuecomment-4128992610). Fixing this would require changes in 3 repos (as of today), so a hard fix unfortunately.


The problem in #26736 is not an exact duplicate of this issue. The root cause of #26736 is that jobs within a reusable workflow do not have corresponding `ActionRunJob` entries on the Gitea side. In #36388, I am attempting to parse the reusable workflow files on the Gitea side and create the missing `ActionRunJob` records so that these jobs can be handled like normal jobs. There are still some bugs I am currently working through.

However, #36388 only provides job-level improvements for reusable workflows - it cannot resolve the other step-related problems in this issue.
